### PR TITLE
Implemented variable inspection when the debugger has started

### PR DESCRIPTION
**Pull request by JohanMabille**
## References

#9114 

## Code changes

This implementation is specific to the debugger package. The frontned now sends an `inspectVariables` request to the backend when the debugger has started and the code has not hit a breakpoint.

This can change can be tested with `xeus-python 0.12.2` or with this [branch of ipykernel](https://github.com/JohanMabille/ipykernel/tree/inspect).

There are still some issues (thus the WIP status):
- infinite loop of sending `inspectVariables` when the user restars the kernel
- defining variables such as dict breaks the behavior (nothing is displayed anymore, but the backend doe snot emit any error).

## User-facing changes

When the debugger has started, the user can now see the variables that have been defined so far in the variables panel of the debugger sidebar.

inspect_variables

## Backwards-incompatible changes

The kernel must support the `inspectVariable` request. This means upgrading to xeus-python 0.12.3 in the tests (this can be done in another PR). There is no impact on ipykernel since the debugger will be supported in the next release only, and the support for `inspectVariables` has been implemented.

### CODE REVIEW

**Comment by JohanMabille** [comment]
xeus-python 0.12.3 is not available yet for all platforms on conda-forge, thus the failure.

**Comment by jtpio** [comment]
Looks like the new wheels will have to be published to PyPI, since xeus-python is installed via pip:

https://github.com/jupyterlab/jupyterlab/blob/ec84e6bb538fbf571c7f583548768e980d738e4d/scripts/ci_install.sh#L52-L55

In the meantime the PR can be tested on Binder: https://mybinder.org/v2/gh/JohanMabille/jupyterlab/inspect?urlpath=lab-dev

File: packages/debugger/src/tokens.ts [hunk 1]
```diff
@@ -80,6 +80,11 @@ export interface IDebugger {
     variablesReference: number
   ): Promise<DebugProtocol.Variable[]>;
 
+  /**
+   *
+   */
```
Comment (jtpio) [inline packages/debugger/src/tokens.ts:85]: Let's add docstrings here and in the other file too.

**Comment by JohanMabille** [comment]
@jtpio I've changed the approach here: a new signal handler is connected to the iopubMessage signal, and asked for the defined variables after an idel message whose parent message type is `execute_request` has been received.

I'm afraid we may miss updates of variables due to widgets, but this avoids storing an internal state.

**Comment by JohanMabille** [comment]
Back to old version of xeus-python, since the upgrade makes the tests fail. We can upgrade and fix the tests in a dedicated PR.

**Comment by JohanMabille** [comment]
I have commented out the "restart" test since it now requires xeus-python 0.12.3. However, upgrading xeus requires fixing other tests. I will upgrade and fix all the tests in another PR.

File: packages/services/test/kernel/ikernel.spec.ts [hunk 2]
```diff
@@ -660,7 +660,8 @@ describe('Kernel.IKernel', () => {
       await expectFailure(restart);
     });
 
-    it('should dispose of existing comm and future objects', async () => {
+    // TODO: enable this test when the CI has upgraded to xeus-python 0.12.3
```
Comment (jtpio) [inline packages/services/test/kernel/ikernel.spec.ts:663]: Just to double check: is this test constantly failing, or was maybe just flaky?

Comment (JohanMabille) [reply packages/services/test/kernel/ikernel.spec.ts]: I've added the test back, let's see.

Comment (JohanMabille) [reply packages/services/test/kernel/ikernel.spec.ts]: Now passing, so my assumption was wrong. I will rebase and drop the last two commits.

**Comment by blink1073** [comment]
Super green!

**Review by blink1073 — APPROVED** [review]
Thanks!

**Comment by lawrence-z-arabii** [comment]
I suppose it's off topic but any ideas why 'table view' is only partly of variable's window height and has its own scrollbar?

[jupyterlab 3.0.14]

image

**Comment by jtpio** [comment]
@lawrence-z-arabii do you mind opening a separate issue to track this? (if none already exists)

The variable inspection from this PR is not in `3.0.14` so it's probably not directly related to this change. But it is in `3.1.0a5` if you want to try it out.

Thanks!

**Pull request merged**